File: design_patterns/C++/Creational/prototype.cxx
#include <iostream>
#include <string>

class Sheep {
  protected:
    std::string name = "";
    std::string category = "";

  public:
    Sheep(std::string name, std::string category = "Mountain Sheep") {
        this->name = name;
        this->category = category;
    }

    Sheep(Sheep const& object) {}

    void setName(std::string name) {
        this->name = name;
    }

    std::string getName() {
        return this->name;
    }

    void setCategory(std::string category) {
        this->category = category;
    }

    std::string getCategory() {
        return this->category;
    }
};

int main() {
    auto original = new Sheep("Jolly");
    std::cout << original->getName() << std::endl;
    std::cout << original->getCategory() << std::endl;
    // ==============
    // Jolly
    // Mountain Sheep
    // ==============
    auto cloned = original;
    cloned->setName("Dolly");
    std::cout << cloned->getName() << std::endl;
    std::cout << cloned->getCategory() << std::endl;
    // ==============
    // Dolly
    // Mountain Sheep
    // ==============
    return 0;
}

File: design_patterns/Python/Creational/prototype.py
import copy


class Sheep:

    def __init__(self, name, category='Mountain Sheep'):
        self.name = name
        self.category = category

    def set_name(self, name):
        self.name = name

    def get_name(self):
        return self.name

    def set_category(self, category):
        self.category = category

    def get_category(self):
        return self.category


if __name__ == '__main__':

    original = Sheep('Jolly')
    assert original.get_name() == 'Jolly'
    assert original.get_category() == 'Mountain Sheep'

    cloned = copy.deepcopy(original)
    cloned.set_name('Dolly')
    assert cloned.get_name() == 'Dolly'
    assert cloned.get_category() == 'Mountain Sheep'
